lib/posix-fdtab: Support independent (per-process) fdtabs

andreittr opened the pull request:
### Description of changes

This PR changes the way `posix-fdtab` handles tables, from a single, static, OS-wide table, to a thread-local reference to one of possibly many independent tables. This allows for, among others, per-process independent fdtabs. Specifically, the commits in this changeset:
- Make the currently active fdtab a thread-local reference
- Support dynamically allocated (and freed) fdtabs
- Handle the `clone()` syscall to duplicate the fdtable when needed

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

`CONFIG_LIBPOSIX_FDTAB=y`

All single-process code should keep working as-is, with a shared fdtab (albeit through independent, refcounted references).
Once multi-process support is added, every thread spawned with `clone()` without the `CLONE_FILES` flag should have independent tables.

andreittr committed:
lib/posix-fdtab: Support independent fdtab refs

This change adds the config option LIBPOSIX_FDTAB_MULTITAB that enables
independent per-thread fdtab references. If enabled, each thread will
lookup an fdtab reference in its TLS memory.
The static initial fdtab remains, and continues to be used.

To this end, fdtabs are now refcounted and have non-static lifetimes.
New threads inherit their parent's ref, with other components or
their callbacks responsible for initializing a meaningful value.
The init thread, however, holds a reference to the static init fdtab.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/posix-fdtab: Add value assert in fmap_xchg

This change adds an assert to sanity-check the value to be exchanged
into the fmap. Exchanging in a NULL value is an invalid operation and
breaks the semantics of fmap, leaving the data structure in a
potentially unsound state.
Calling code should never normally do this; the assert serves as extra
precaution for future development.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/posix-fdtab: Handle fdtab duplication on clone

This change adds logic to handle fdtab references on clone:
- if  CLONE_FILES: child inherits a reference to parent's tab
- if !CLONE_FILES: child allocates new fdtab duplicate of parent's,
  populated with new references to the same open file descriptions hosted
  by the parent fdtab.

The initial duplication logic is rudimentary and does not provide any
ordering guarantees w.r.t. syscalls modifying the original table (open,
dup, close), under the assumption that it won't trigger race conditions
in the wild. Please revisit if this turns out to be overly optimistic.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/posix-fdtab: Simplify active fdtab lookup

This change refactors the way fdtab code looks up the currently active
fdtable, simplifying functions and removing boilerplate code.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Rebased to current staging.

John-Ted reviewed (approved):
Looks good to me

Reviewed-by: Ioan-Teodor Teugea ioan_teodor.teugea@stud.acs.upb.ro

csvancea reviewed (approved):
Reviewed-by: Cosmin Vancea csvancea@gmail.com

andreittr commented:
Rebased to current staging

michpappas commented on lib/posix-fdtab/fdtab.c:
```diff
@@ -432,6 +437,63 @@ static void fdtab_thread_term(struct uk_thread *child)
+		struct uk_fdtab *newtab = fdtab_duplicate(tab);
+
+		if (unlikely(!newtab))
+			return -ENOMEM;
```
Very minor: In case `fdtab_duplicate()` fails for additional reasons in the future, instead of making assumptions on  the caller's side, consider if it would be preferable to update `fdtab_duplicate()` to return error pointer that the caller can directly decode and return. Alternatively you can consider adding a line in the description "Returns pointer on success or NULL if out of memory".

andreittr replied:
I see your point, though I prefer that functions who have a single possible failure state by design do not return error codes, but let the caller in higher levels interpret their failure. Returning errptrs is particularly tricky because then the caller must worry _both_ about errptrs _and_ NULLs (because NULL is _not_ an error pointer as it doesn't map to a negative errno).

`fdtab_duplicate()` falls into this single fail state group, as it's a glorified malloc + memcpy by design (and the latter cannot fail recoverably); `fdtab_clone` is the higher-level "public-facing" function that is expected to conform to the errno API, and there is IMO the best place to encode and return errors.
I'll update description to make its return values clear.

michpappas replied:
Up to you, especially as its not a public API it's not that critical. Just sharing my 2c.

michpappas commented on lib/posix-fdtab/fdtab.c:
```diff
@@ -432,6 +437,63 @@ static void fdtab_thread_term(struct uk_thread *child)
+{
+	struct uk_fdtab *ret = uk_malloc(tab->alloc, sizeof(*ret));
+
+	if (likely(ret)) {
```
Generally speaking, the convention in the codebase is to wrap checks for errors with `unlikely` and return immediately, and not use `likely` in general. The reason is more than style: If a block wrapped around `likely()` calls an inline function that involves error handling, that function will almost certainly use `unlikely`, and that results into multiple likelihood attributes in the same execution path, which can result into the [opposite than the desired effect](https://blog.aaronballman.com/2020/08/dont-use-the-likely-or-unlikely-attributes/).

andreittr replied:
I also tend to use `unlikely` in general and keep `likely` for only this exact pattern: "call, likely init, ret", as that's "safe" from unintended effects in isolation. But I hadn't considered the effects of inlining or LTO before :thinking: good point; will replace with unlikely.

michpappas commented on lib/posix-fdtab/fdtab.c:
```diff
@@ -432,6 +437,63 @@ static void fdtab_thread_term(struct uk_thread *child)
+	return 0;
+}
+
+UK_POSIX_CLONE_HANDLER(CLONE_FILES, 0, fdtab_clone, 0);
```
Not necessarily in the scope of this PR, but from the documentation for `CLONE_FILES` in the man page, it looks like we need to update the `fdtab_handle_execve()` to handle the case where the fdtab is shared between the parent and the child:
> If a process sharing a file descriptor table calls execve(2), its file descriptor table is duplicated (unshared).

andreittr replied:
I agree we have to update fdtab for `execve()`, and more generally make "fdtab unsharing" a singular well-defined event that can be triggered either by `clone()` or `execve()`, and can in turn trigger further actions.
I believe we need this in order to correctly support "magic" files that are unshared implicitly along with the fdtab, without explicit action from userspace, such as `epollfd` and `signalfd`.

michpappas replied:
Alright, thanks for confirming. Let's treat it as a follow up item then 👍🏼

michpappas commented on lib/posix-fdtab/fmap.h:
```diff
@@ -184,6 +184,23 @@ static inline void uk_fmap_init(const struct uk_fmap *m)
+ */
+static inline void uk_fmap_set(const struct uk_fmap *m, int idx, const void *p)
+{
+	if (!_FMAP_INRANGE(m, idx))
```
Just curious cause I notice that you make the same check in other functions in this header: is the out-of-range an expected case that we can safely continue execution from, and therefore not qualifying for an assertion?

andreittr replied:
:thinking: fdtab ops/syscalls _should_ return -EBADF if called with an out-of-bounds fd, so that's explicitly a use-case we need to safely continue from, and from looking at the code it seems fdtab relies on fmap to signal errors instead of bounds-checking at higher levels.

I see your point that a cleaner design would do bounds checking & `return -EBADF` in the fdtab ops/syscalls themselves and have the raw internal data structure at most assert bounds -- separation of mechanism and policy and all that. While such a refactor is not something I'm comfortable tacking on to this PR, I think the broader VFS work (i.e., now-ish) is a good time to do this.

michpappas replied:
Yes of course it's out of scope. Just a general observation as I was going through the code 👍🏼

andreittr commented:
Rebased to current staging

andreittr commented:
Addressed comments

andreittr commented:
Rebased to current staging

andreittr commented:
Pushed a minor rework; in short:
- single fdtab or multi-fdtab is now a hidden Kconfig option, autoselected if `clone()` support is enabled
  - this neatly decouples the mechanism of "having multiple fdtabs" from the policy of "what fdtab each thread has"
  - also: if no `clone()`, no indirection => less overhead for single-thread apps
- newly-created threads no longer inherit their parent's fdtab reference; raw `uk_thread`s get NULL (kernel threads shouldn't be mucking about with the fdtab anyway), `clone` handler implements the semantics of sharing/unsharing for posix threads
- the init thread still starts with a reference to the init fdtab; this is in lieu of a "thread posixification" callback, and once such a mechanism exists, we should probably move the init fdtab there

michpappas reviewed (approved):
As discussed: once #1364 is merged the `POSIX_PROCESS_EXIT_EVENT` will be available to terminate the fdtab of posix threads.

Other than that...

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

andreittr commented:
Pushed a compat stop-gap change to have all raw threads inherit the fdtab ref of their parent (instead of NULL).

andreittr committed:
lib/posix-fdtab: Fix wrong return in exec handler

Previously fdtab_handle_execve would return 0 on success, as per common
convention. This is however wrong for event handlers, as these require
specific exit codes on success; in this case UK_EVENT_HANDLED_CONT.
This change fixes this oversight.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Pushed a tiny bonus fix for the wrong value being returned by the `execve` event handler.

andreittr commented:
Pushed a fix for a type mismatch that resulted in misaligned accesses on arm64.

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
